peer/peer_test.go

```diff
@@ -0,0 +1,362 @@
+	p.RemoveTxMsgListener("handleTxMsg")
+	p.RemoveBlockMsgListener("handleBlockMsg")
+	p.RemoveInvMsgListener("handleInvMsg")
+	p.RemoveHeadersMsgListener("handleHeadersMsg")
+	p.RemoveGetDataMsgListener("handleGetDataMsg")
+	p.RemoveGetBlocksMsgListener("handleGetBlocksMsg")
+	p.RemoveGetHeadersMsgListener("handleGetHeadersMsg")
+	p.RemoveGetAddrMsgListener("handleGetAddrMsg")
+	p.RemoveAddrMsgListener("handleAddrMsg")
+}
+
+// TestPeerConnection tests the activity between inbound and outbound peers
+// using a mock connection.
+func TestPeerConnection(t *testing.T) {
+	peerCfg := &peer.Config{
+		UserAgentName:    "peer",
+		UserAgentVersion: "1.0",
+	}
+	c1, c2 := pipe(
+		&conn{raddr: "127.0.0.1:8333"},
+		&conn{raddr: "127.0.0.1:54833"},
+	)
+
+	p1 := peer.NewInboundPeer(newestSha, 0, wire.MainNet, wire.SFNodeNetwork, addrMgr, peerCfg, c1)
+	registerListeners(p1)
+	err := p1.Start()
+	if err != nil {
+		t.Errorf("Start: error %v", err)
+		return
+	}
+
+	p2 := peer.NewOutboundPeer(newestSha, 1, wire.MainNet, wire.SFNodeNetwork, addrMgr, peerCfg, c2, "127.0.0.1:8333")
+
+	locator := blockchain.BlockLocator([]*wire.ShaHash{})
+	p2.PushGetBlocksMsg(locator, &zeroHash)
+	p2.PushGetHeadersMsg(locator, &zeroHash)
+	p2.PushRejectMsg(wire.CmdBlock, wire.RejectInvalid, "test reject", &zeroHash, false)
+
+	fakeHeader := wire.NewBlockHeader(&zeroHash, &zeroHash, 1, 1)
+	msgBlock := wire.NewMsgBlock(fakeHeader)
+
+	// Test protocol messages
+	p2.QueueMessage(wire.NewMsgPing(uint64(p1.ID())), nil)
+	p2.QueueMessage(wire.NewMsgGetAddr(), nil)
+	p2.QueueMessage(msgBlock, nil)
+	p2.QueueMessage(wire.NewMsgMemPool(), nil)
+	p2.QueueMessage(wire.NewMsgTx(), nil)
+	p2.QueueMessage(wire.NewMsgInv(), nil)
+	p2.QueueMessage(wire.NewMsgHeaders(), nil)
+	p2.QueueMessage(wire.NewMsgGetData(), nil)
+	p2.QueueMessage(wire.NewMsgGetBlocks(&zeroHash), nil)
+	p2.QueueMessage(wire.NewMsgGetHeaders(), nil)
+	p2.QueueMessage(wire.NewMsgAddr(), nil)
+
+	// Test adding block inv
+	fakeInv := wire.NewInvVect(wire.InvTypeBlock, &zeroHash)
+	p2.QueueInventory(fakeInv)
+
+	removeListeners(p2)
+
+	// Test all handlers are called
+out:
+	for {
+		select {
+		case handler := <-gotHandlers:
+			delete(wantHandlers, handler)
+			if len(wantHandlers) == 0 {
+				break out
+			}
+		default:
+		}
+	}
+
+	// Test peer flags and stats
+	testPeer(t, p1, true)
+	testPeer(t, p2, false)
+}
+
+// testPeer tests the given peer's flags and stats
+func testPeer(t *testing.T, p *peer.Peer, inbound bool) {
+	wantID := int32(1)
+	wantAddr := "127.0.0.1:8333"
+	wantUserAgent := "/btcwire:0.2.0/peer:1.0/"
+	wantInbound := true
+	wantServices := wire.SFNodeNetwork
+	wantProtocolVersion := uint32(70002)
+	wantConnected := true
+	wantVersionKnown := true
+	wantLastBlock := int32(234439)
+	wantStartingHeight := int32(234439)
+	wantLastPingNonce := uint64(0)
+	wantLastPingMicros := int64(0)
+	wantTimeOffset := int64(0)
+
+	if !inbound {
+		wantID = 2
+		wantAddr = "127.0.0.1:8333"
```

Comment:
This and several of the other fields are already set to the same values above.  No need to override them to the exact same values.